src/System.Diagnostics.Process/tests/ProcessTests.cs

```diff
@@ -218,6 +218,89 @@ public void ProcessStart_UseShellExecute_OnWindows_DoesNotThrow(bool isFolder)
+            if (PlatformDetection.IsWindows)
+            {
+                filename += ".bat";
+            }
+            WriteScriptFile(filename, returnValue: 42);
+
+            if (filenameAsUrl)
+            {
+                filename = new Uri(filename).ToString();
+            }
+
+            using (var process = Process.Start(new ProcessStartInfo { UseShellExecute = true, FileName = filename }))
+            {
+                process.WaitForExit();
+                Assert.Equal(42, process.ExitCode);
+            }
+        }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsServerCore),
+            nameof(PlatformDetection.IsNotWindowsNanoServer), nameof(PlatformDetection.IsNotWindowsIoTCore))]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap, "not supported on UAP")]
+        public void ProcessStart_UseShellExecute_ExecuteOrder()
+        {
+            // Create a directory that we will use as PATH
+            string path = GetTestFileName();
+            Directory.CreateDirectory(path);
+
+            RemoteInvokeOptions options = new RemoteInvokeOptions();
+            options.StartInfo.EnvironmentVariables["PATH"] = path;
+            RemoteInvoke(pathDirectory =>
+            {
+                // Create two identically named scripts, one in the working directory and one on PATH.
+                string scriptFilename = $"script_{nameof(ProcessStart_UseShellExecute_ExecuteOrder)}";
+                if (PlatformDetection.IsWindows)
+                {
+                    scriptFilename += ".bat";
+                }
+                const int workingDirReturnValue = 1;
+                const int pathDirReturnValue = 2;
+                WriteScriptFile(Path.Combine(pathDirectory, scriptFilename), returnValue: pathDirReturnValue);
+                WriteScriptFile(Path.Combine(Directory.GetCurrentDirectory(), scriptFilename), returnValue: workingDirReturnValue);
+
+                // Execute the script and verify we prefer the one in the working directory.
+                using (var process = Process.Start(new ProcessStartInfo { UseShellExecute = true, FileName = scriptFilename }))
+                {
+                    process.WaitForExit();
+                    Assert.Equal(workingDirReturnValue, process.ExitCode);
+                }
+
+                // Remove the script in the working directory and verify we now use the one on PATH.
+                File.Delete(scriptFilename);
+                using (var process = Process.Start(new ProcessStartInfo { UseShellExecute = true, FileName = scriptFilename }))
+                {
+                    process.WaitForExit();
+                    Assert.Equal(pathDirReturnValue, process.ExitCode);
+                }
+
+                return SuccessExitCode;
+            }, path, options).Dispose();
+        }
+
+        private void WriteScriptFile(string filename, int returnValue)
```

Comment:
this could return a file name and encapsulate logic for different platforms (and also use temporary file)